src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt

```diff
@@ -270,14 +280,18 @@ class RsInferenceContext(
             val impl = lookup.select(resolveTypeVarsIfPossible(traitRef)).ok()?.impl as? RsImplItem ?: continue
             val fn = impl.expandedMembers.functions.find { it.name == fnName } ?: continue
             val source = TraitImplSource.ExplicitImpl(RsCachedImplItem.forImpl(impl))
-            resolvedPaths[path] = listOf(ResolvedPath.AssocItem(fn, source))
+            val result = ResolvedPath.AssocItem(fn, source)
+            result.subst = variant.subst // TODO remap subst
+            resolvedPaths[path] = listOf(result)
         }
         for ((call, traitRef) in methodRefinements) {
-            val variant = resolvedMethods[call]?.firstOrNull() ?: continue
+            val info = resolvedMethods[call] ?: continue
+            val variant = info.resolveVariants.firstOrNull() ?: continue
             val impl = lookup.select(resolveTypeVarsIfPossible(traitRef)).ok()?.impl as? RsImplItem ?: continue
             val fn = impl.expandedMembers.functions.find { it.name == variant.name } ?: continue
             val source = TraitImplSource.ExplicitImpl(RsCachedImplItem.forImpl(impl))
-            resolvedMethods[call] = listOf(variant.copy(element = fn, source = source))
+            // TODO remap subst
```

Comment:
Note that this doesn't work correctly for refined paths (see test infer method arg with multiple impls of the same trait or test resolve UFCS method call with multiple impls of the same trait). I'll fix it later